src/game/common/statemachine.cpp

```diff
@@ -0,0 +1,716 @@
+}
+
+Object *StateMachine::Get_Goal_Object()
+{
+    return g_theGameLogic->Find_Object_By_ID(m_goalObjectID);
+}
+
+Object const *StateMachine::Get_Goal_Object() const
+{
+    return g_theGameLogic->Find_Object_By_ID(m_goalObjectID);
+}
+
+void StateMachine::Set_Goal_Position(Coord3D const *pos)
+{
+    if (!m_locked) {
+        Internal_Set_Goal_Position(pos);
+    }
+}
+
+void StateMachine::Internal_Set_Goal_Position(Coord3D const *pos)
+{
+    if (pos) {
+        m_goalPosition = *pos;
+    }
+}
+
+void StateMachine::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+    xfer->xferUnsignedInt(&m_updateFrame);
+    xfer->xferUnsignedInt(&m_defaultStateID);
+    unsigned int id = Get_Current_State_ID();
+    xfer->xferUnsignedInt(&id);
+
+    if (xfer->Get_Mode() == XFER_LOAD) {
+        m_currentState = Internal_Get_State(id);
+    }
+
+    bool b = false;
+    xfer->xferBool(&b);
+
+    if (b) {
+        int count = 0;
+
+        for (auto &iter : m_stateMap) {
+            count++;
+        }
+
+        int count2 = count;
+        xfer->xferInt(&count2);
+        captainslog_dbgassert(count2 == count, "State count mismatch - %d expected, %d read", count, count2);
+
+        for (auto &iter : m_stateMap) {
+            State *state = iter.second;
+            unsigned int state_id = state->Get_ID();
+            xfer->xferUnsignedInt(&state_id);
+
+            captainslog_dbgassert(
+                state_id == state->Get_ID(), "State ID mismatch - %d expected, %d read", state->Get_ID(), state_id);
+
+            if (state == nullptr) {
+                captainslog_dbgassert(state != nullptr, "state was NULL on xfer, trying to heal...");
+                state = Internal_Get_State(m_defaultStateID);
+            }
+
+            state->Xfer_Snapshot(xfer);
+        }
+    } else {
+        if (m_currentState == nullptr) {
+            captainslog_dbgassert(m_currentState != nullptr, "currentState was NULL on xfer, trying to heal...");
+            m_currentState = Internal_Get_State(m_defaultStateID);
+        }
+
+        m_currentState->Xfer_Snapshot(xfer);
+    }
+
+    xfer->xferObjectID(&m_goalObjectID);
+    xfer->xferCoord3D(&m_goalPosition);
+    xfer->xferBool(&m_locked);
+    xfer->xferBool(&m_defaultStateInited);
+}
+
+unsigned int StateMachine::Get_Current_State_ID() const
+{
+    if (m_currentState) {
+        return m_currentState->Get_ID();
+    } else {
+        return INVALID_STATE_ID;
+    }
+}
+
+const Coord3D *StateMachine::Get_Goal_Position() const
+{
+    return &m_goalPosition;
+}
+
+Object *StateMachine::Get_Owner()
+{
+    return m_owner;
+}
+
+Object const *StateMachine::Get_Owner() const
+{
+    return m_owner;
+}
+
+bool StateMachine::Is_In_Attack_State() const
+{
+    return m_currentState == nullptr || m_currentState->Is_Attack();
+}
+
+bool StateMachine::Is_In_Guard_Idle_State() const
+{
+    return m_currentState == nullptr || m_currentState->Is_Guard_Idle();

```

Comment:
Hmm. This condition looks very odd. The way its written right now would mean if current state is null, then both Attack State and Guard Idle State are true.

```c++
return m_currentState ? m_currentState->Is_Guard_Idle() : false;
```
would be more logical here.

Reply:
Its how the original code did it.

Reply:
Please add a comment to investigate.

Reply:
I took another look and it turns out Is_In_Guard_Idle_State was wrong so I fixed that but Is_In_Attack_State is correct so there is nothing to investigate.